contrib/kind.sh

```diff
@@ -150,57 +150,22 @@ print_params
 
 set -euxo pipefail
 
-# Detect IP to use as API server
-API_IPV4=""
```

Comment:
How is the API IP now determined?

Reply:
KIND deploys the apiserver, so it knows the IP :smile: . 
It also maps the IP to a port in the host, we can get it from there ( you were using this until now)
```
db1eee5a6b7c        kindest/node:latest   "/usr/local/bin/entr…"   24 hours ago        Up 24 hours         127.0.0.1:39685->6443/tcp   kind-control-plane
```
 
but that means that the traffic is round tripping from the container to the host, natting, ... It's more optimal to get the IP address of the container, so the pods can reach it directly, and we can obtain it from the kubeconfig file

Reply:
@aojea how does KIND know which IP to choose? In your example it is choosing localhost, which the old code explicitly prevented.

Reply:
maybe I'm not understanding well the question, let me explain the whole process.
In Kind there are 2 node roles, control plane and worker nodes, the control plane host the apiserver.
The apiserver needs to configure the api-advertise-adress, that's the endpoint of the API, this is configured here
https://github.com/kubernetes-sigs/kind/blob/5c4a5ad54a8711b69b5f1143ce23f0d3a4bf807d/pkg/cluster/internal/kubeadm/config.go#L268-L272
using the node address (nodes only has one interface)

If there are multiple control planes, KIND creates a load balancer containers to loadbalance the apiservers.

There is a method that returns the apiserver endpoint 
https://github.com/kubernetes-sigs/kind/blob/65a3f36d6d46e5469bc56573d8de68427da7d5b1/pkg/cluster/internal/providers/docker/provider.go#L148-L149
This is used to generate the kubeconfig file, and to put the API endpoint in the `server:` field, that's what we are using now to get the API address. Since, we need this address for pods that are running in the cluster, we used the `--internal` flag, that means is going to give us the address reachable from the containers.

Reply:
So when KIND goes to bind the port that it NAT's to 6443, it chooses an IP address. In your example it is 127.0.0.1. What I'm saying is in the old code, that is explicitly not allowed. So that the IP chosen is a real IP on on your host, which will propagate to your kubeconfig. Meaning that, you can use this kubeconfig outside of your server (say on your laptop) and it will work. Example:
```
db40fa13c0db        kindest/node:v1.18.2   "/usr/local/bin/entr…"   8 hours ago         Up 8 hours          10.8.125.15:11337->6443/tcp   ovn-control-plane
```

[trozet@nfvsdn-05 contrib]$ cat ~/admin.conf |grep 10.8
    server: https://10.8.125.15:11337

Reply:
ahh, ok, got it, we are talking about the "published" ip address.
This is the one that KIND maps to the host doing portmapping, by default is 127.0.0.1, but it's configurable and until now these scripts were configuring it using the host address
https://github.com/ovn-org/ovn-kubernetes/blob/c2578919fa7d217c0cea5a2933eacd7941cebea2/contrib/kind.yaml.j2#L6

Reply:
So based off that, sounds like I need to back this change out. So we can access the local clusters (not necessarily Github CI clusters) from a remote server.

Reply:
that's a security hole that we are trying to convince people not to use :shrug:

Reply:
IMO this isn't meant for production it's a development/test environment. There is no security here, for example we don't use SSL/TLS on the OVN/OVS connections. Regardless, we dont need to change that behavior in this PR. We can open another PR and give the option of using localhost or not if some folks do not feel comfortable opening up a port on their deployment host. I would suggest backing this out for now, since it isn't required for IPv6 and opening a new PR on it.

Reply:
There are different things here:
- the security is because you are exposing your APIserver to the public, communication inside the kind cluster are running inside a host.
- if we want to keep exposing the apiserver so users can connect to the deployment from outside, there is no need to find the host address, just use `apiServerAddress: 0.0.0.0` and it will listen in all interfaces
- pods must use the internal IP address of the API server, this avoid possible issues with NAT inside the host  Instead of podA -> hostIP:PortMapped -> PodAPIserver:PortAPI we go directly